haskell/private/path_utils.bzl

```diff
@@ -318,3 +318,103 @@ def copy_all(ctx, srcs, dest):
             command = "mkdir -p {dest} && cp -L -R \"$@\" {dest}".format(dest = dest.path),
             arguments = [args],
         )
+
+def parse_pattern(pattern_str):
+    """Parses a string label pattern.
+
+    Args:
+      pattern_str: The pattern to parse.
+        Patterns are absolute labels in the local workspace. E.g.
+        `//some/package:some_target`. The following wild-cards are allowed:
+        `...`, `:all`, and `:*`. Also the `//some/package` shortcut is allowed.
+
+    Returns:
+      A struct of
+        package: A list of package path components. May end on the wildcard `...`.
+        target: The target name. None if the package ends on `...`. May be one
+          of the wildcards `all` or `*`.
+
+    """
+
```

Comment:
Should you start by trimming leading and trailing whitespaces from the pattern and the needle? I'm not sure, but this may be discussed.